fv3core/stencils/a2b_ord4.py

```diff
@@ -22,485 +36,435 @@
+            qx = qx_edge_west2(qin, dxa, qx)
+
+        with horizontal(region[i_end + 1, :]):
+            qx = qx_edge_east(qin, dxa)
+        with horizontal(region[i_end, :]):
+            qx = qx_edge_east2(qin, dxa, qx)
+
+        # compute_qy
+        qy = ppm_volume_mean_y(qin)
+        with horizontal(region[:, j_start]):
+            qy = qy_edge_south(qin, dya)
+        with horizontal(region[:, j_start + 1]):
+            qy = qy_edge_south2(qin, dya, qy)
+
+        with horizontal(region[:, j_end + 1]):
+            qy = qy_edge_north(qin, dya)
+        with horizontal(region[:, j_end]):
+            qy = qy_edge_north2(qin, dya, qy)
+
+        # compute_qxx
+        qxx = lagrange_y(qx)
+        with horizontal(region[:, j_start + 1]):
+            qxx = cubic_interpolation_south(qx, qout, qxx)
+        with horizontal(region[:, j_end]):
+            qxx = cubic_interpolation_north(qx, qout, qxx)
+
+        # compute_qyy
+        qyy = lagrange_x(qy)
+        with horizontal(region[i_start + 1, :]):
+            qyy = cubic_interpolation_west(qy, qout, qyy)
+        with horizontal(region[i_end, :]):
+            qyy = cubic_interpolation_east(qy, qout, qyy)
+
+        with horizontal(region[i_start + 1 : i_end + 1, j_start + 1 : j_end + 1]):
+            qout = 0.5 * (qxx + qyy)
+        # }
+
+        if __INLINED(REPLACE):
+            qin = qout
+
+
+def _make_grid_storage_2d(grid_array: gt4py.storage.Storage, index: int = 0):
+    grid = spec.grid
+    return gt4py.storage.from_array(
+        grid_array[:, :, index],
+        backend=global_config.get_backend(),
+        default_origin=grid.compute_origin()[:-1],
+        shape=grid_array[:, :, index].shape,
+        dtype=grid_array.dtype,
+        mask=(True, True, False),
     )
 
 
-def compute(qin, qout, kstart=0, nk=None, replace=False):
+def compute(
+    qin: FloatField,
+    qout: FloatField,
+    kstart: int = 0,
+    nk: Optional[int] = None,
+    replace: bool = False,
+):
+    grid = spec.grid
     if nk is None:
-        nk = grid().npz - kstart
-    extrapolate_corners(qin, qout, kstart, nk)
-    if spec.namelist.grid_type < 3:
-        compute_qout_edges(qin, qout, kstart, nk)
-        qx = compute_qx(qin, qout, kstart, nk)
-        qy = compute_qy(qin, qout, kstart, nk)
-        qxx = compute_qxx(qx, qout, kstart, nk)
-        qyy = compute_qyy(qy, qout, kstart, nk)
-        compute_qout(qxx, qyy, qout, kstart, nk)
-        if replace:
-            copy_stencil(
-                qout,
-                qin,
-                origin=(grid().is_, grid().js, kstart),
-                domain=(grid().ie - grid().is_ + 2, grid().je - grid().js + 2, nk),
-            )
-    else:
-        raise Exception("grid_type >= 3 is not implemented")
+        nk = grid.npz - kstart
+    agrid1 = _make_grid_storage_2d(grid.agrid1)
+    agrid2 = _make_grid_storage_2d(grid.agrid2)
+    bgrid1 = _make_grid_storage_2d(grid.bgrid1)
+    bgrid2 = _make_grid_storage_2d(grid.bgrid2)
```

Comment:
So these could all be temporary storages inside the stencil, but we can't have 2d storages, so I can see why you'd keep it outside the stencil call. But what is worse, allocating new 2d storages every timestep and passing them, or having the stencil allocate 3d storages on the fly but taking advantage of the 'wheel' (or will the 'wheel' be available in this scope too? my impression was no?).

Reply:
These should be just views on the 3D data for the sake of making the stencils using `FloatFieldIJ` in the gtscript. This shouldn't have a large impact, and can be removed when the grid storage refactor is complete.

Reply:
Ah gotchya, thanks for the clarification